File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.cpp
class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& points) {
    //    minimum spanning tree- all the points should be connected with minimum weight
    // make a adjlist  from all the points , one node can be connected to rest of all the nodes
    // int n=points.size();
    // unordered_map<int, vector<pair<int,int>>>adj;
    // for(int i=0;i<n;i++){
    //     for(int j=i+1;j<n; j++){
    //         int distance= abs(points[i][0]-points[j][0])+abs(points[i][1]-points[j][1]);
    //         adj[i].push_back({j,distance});
    //         adj[j].push_back({i,distance});
    //     }
    // }
    // priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> pq;
    // int minCost=0;
    // pq.push({0,0});
    // vector<int>visited(n,0);
    // while(!pq.empty()){
    //     auto it= pq.top();
    //     int node=it.second;
    //     int wt=it.first;
    //     pq.pop();
    //     if(visited[node]) continue;
    //     visited[node]=1;
    //     minCost+=wt;
    //     for(auto it: adj[node]){
    //         int adjNode=it.first;
    //         int edgeWt=it.second;
    //         if(!visited[adjNode]){
    //             pq.push({edgeWt,adjNode});
    //         }
    //     }
    // }
    // return minCost;


    int n=points.size();
    priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> pq;
    pq.push({0,0});
    unordered_map<int,int>cache;
    int min_cost=0;
vector<int>visited(n,0);
    while(!pq.empty()){
        auto [wt,node]=pq.top();
        pq.pop();

if(visited[node]) continue;
visited[node]=1;
min_cost+=wt;
for (int i=0;i<n;i++){ //for ever current node, try to check from 0 to nth node
    if(!visited[i]){
    int dist=abs(points[node][0]-points[i][0])+abs(points[node][1]-points[i][1]);
// agr distance node se i tkk ka less hai than stored distance to reach i in cache , update the cache
if(cache.find(i)==cache.end() || dist<cache[i]){
    cache[i]=dist;
    pq.push({dist, i});
}
    }
}
    }
return min_cost;
    }
};

File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.js
/**
 * @param {number[][]} points
 * @return {number}
 */
var minCostConnectPoints = function(points) {
//      const n=points.length;
//    const adj=new Map();
//    for(let i=0;i<n;i++){
//             adj.set(i,[]);
//     }
//     for(let i=0;i<n;i++){
//         for(let j=i+1;j<n; j++){
//             const distance= Math.abs(points[i][0]-points[j][0])+Math.abs(points[i][1]-points[j][1]);
//             adj.get(i).push([j,distance]);
//             adj.get(j).push([i,distance]);
//         }
//     }
//     const pq=[];
//     let minCost=0;
//     pq.push([0,0]);
//     const visited=Array(n).fill(0);
//     while(pq.length>0){
//         pq.sort((a,b)=>a[0]-b[0]);//sort based on first parameter
//        const [wt,node]= pq.shift();
       
//         if(visited[node]) continue;
//         visited[node]=1;
//         minCost+=wt;
//         for(const [adjNode,edgeWt] of  adj.get(node)){
           
//             if(!visited[adjNode]){
//                 pq.push([edgeWt,adjNode]);
//             }
//         }
//     }
//     return minCost;


// remove pq completely as it adds extra time complexity 
const n=points.length;
const visited= Array(n).fill(false);
const distance=Array(n).fill(Infinity);//stores min distance to reach node i

distance[0]=0;
let min_cost=0;

for(let i=0;i<n;i++){
    let u=getMinDistanceVertex(visited,distance);

    visited[u]=true;
    min_cost+=distance[u];

    for(let j=0;j<n;j++){
        if(!visited[j]){
            const dist=Math.abs(points[u][0]-points[j][0])+Math.abs(points[u][1]-points[j][1]);
            distance[j]=Math.min(distance[j],dist);
        }
    }

}
return min_cost;
};

var getMinDistanceVertex=function(visited,distance){
    let minVertex=-1;
    let minDist=Infinity;
    for(let i=0;i<distance.length;i++){
        if(!visited[i] && distance[i]<minDist){
            minDist=distance[i];
            minVertex=i;
        }
    }
    return minVertex;
}